src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/UrlEncoder.cs

```diff
@@ -177,6 +175,12 @@ public unsafe override bool TryEncodeUnicodeScalar(int unicodeScalar, char* buff
                     numberOfCharactersWritten = 0;
                     return false;
                 }
+
+                if (numberOfCharactersWritten + 3 > bufferLength)
+                {
+                    throw new ArgumentException("Argument encoder does not implement MaxOutputCharsPerInputChar correctly.");
+                }
```

Comment:
This is a public API that has nothing to do with MaxOutputCharsPerInput. I don't think it should throw this exception. It should do what the buggy check before did: return false (meaning the buffer was too small; that's why it's a Try* method). MaxOutputCharsPerInput is just a hint; a performance optimization to avoid buffer resizing and copies.